### DISCUSSION TITLE
.NETCore Kestrel HTTPs 502 Bad Gateway nginx with port forwarding

### DISCUSSION THREAD

**Discussion by warrenbuckley** [community/community#7116] [Codespaces]
## Summary
Unable to browse to a dotnet new WebApp template on HTTPs 5001 port. A 502 Bad Gateway error is shown from nginx

![image](https://user-images.githubusercontent.com/1389894/139735531-900808a3-fca8-42b1-ac35-c68dc26d5f77.png)

## Terminal Output
```
@warrenbuckley ➜ /workspaces/CodeSpace-NetCore-HTTPS/testing (main ✗) $ dotnet run
Building...
warn: Microsoft.AspNetCore.DataProtection.Repositories.FileSystemXmlRepository[60]
      Storing keys in a directory '/root/.aspnet/DataProtection-Keys' that may not be persisted outside of the container. Protected data will be unavailable when container is destroyed.
warn: Microsoft.AspNetCore.Server.Kestrel[0]
      Unable to bind to https://localhost:5001 on the IPv6 loopback interface: 'Cannot assign requested address'.
warn: Microsoft.AspNetCore.Server.Kestrel[0]
      Unable to bind to http://localhost:5000 on the IPv6 loopback interface: 'Cannot assign requested address'.
info: Microsoft.Hosting.Lifetime[0]
      Now listening on: https://localhost:5001
info: Microsoft.Hosting.Lifetime[0]
      Now listening on: http://localhost:5000
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Development
info: Microsoft.Hosting.Lifetime[0]
      Content root path: /workspaces/CodeSpace-NetCore-HTTPS/testing
```
![Terminal Output Screenshot](https://user-images.githubusercontent.com/1389894/139734639-0e78269a-ce23-4c0b-8665-c4e0e551dffd.png)

## Steps to Reproduce:

1. Create a new blank repository with readme
2. Open new repo in Github Codespaces
3. Add DevContainer configuration for .NET C# with MSSQL
4. Add Github CLI & Azure CLI as optional extras
5. From terminal `dotnet new webapp --name Testing`
6. `cd Testing` and then `dotnet restore`
7. `dotnet build` and `dotnet run`
8. See error/warning about missing local HTTPS certificates
```
Microsoft.AspNetCore.Server.Kestrel[0]
      Unable to start Kestrel.
      System.InvalidOperationException: Unable to configure HTTPS endpoint. No server certificate was specified, and the default developer certificate could not be found or is out of date.
      To generate a developer certificate run 'dotnet dev-certs https'. To trust the certificate (Windows and macOS only) run 'dotnet dev-certs https --trust'.
```
9. Run `dotnet dev-certs https --trust`
10. Rerun `dotnet-run`
11. Wait for notification prompt about Auto Port Forwarding for 5000 and 5001
12. Try to browse to the HTTPS url which uses port 5001
13. 502 Bad Gateway error from nginx (see screenshot)

![image](https://user-images.githubusercontent.com/1389894/139735531-900808a3-fca8-42b1-ac35-c68dc26d5f77.png)

**Comment by Chuxel** [comment]
I mentioned this in another location you logged this, but try this using VS Code client instead of the web. I don't believe that web port forwarding would support a custom SSL certificate (unless @derekbekoe or @lostintangent correct me).  When using VS Code client, on your local machine this would be mapped to localhost as a straight TCP port and would then work.

You would also have to configure your local browser with the SSL certificate since dotnet new generates a custom one.  Generally a better approach is to create one locally and copy it into a Codespace instead (e.g. as described at https://github.com/microsoft/vscode-dev-containers/tree/main/containers/dotnet#enabling-https-in-aspnet)

EDIT: While forwarding a custom cert is not supported, the web URI should switch from the custom cert to a codespaces one if the port is set in `https` mode. There is a bug with this right now.

**Comment by warrenbuckley** [comment]
Hi @Chuxel /cc @lostintangent & @derekbekoe
Thanks for this info. I have done as you suggested and used VSCode locally on my machine connected to the GitHub Codespace and this works locally. 

## Potential bug ?
With the exception that the 5001 port in the ports pane from Codespaces in VSCode is mapped to http and not https.

In the Ports panel from Codespaces VSCode extension that lists out the ports, the privacy and process etc, I have then tried to set the port 5001 to use HTTPS Port Protocol, but instead this seems to use the `5001.githubpreview.dev` domain and then this does not work if you try to browse to the site from this approach.

Browsing to it directly with CTRL + Click from the terminal from the output of dotnet run command works with no issues as long as you have not changed the Port protocol from the ports pane.

## Questions
With me doing the steps for the development SSL certificate from my machine into the Codespace. If any colleagues, team mates or anyone else who would fork or work on this repo. Would the certificate work for them without issues or does each new person working on the repository need to repeat these steps?

I am planning on creating a GitHub Repository template, with the goal for someone to have a specific .NETCore project ready for them to go and work with direct in GitHub Codespaces. How could I automate the SSL certificate export so its one less step for people to worry about to get up and running.

Are there any plans to support the `githubpreview.dev` for HTTPS as it would make a nicer experience to allow developing direct in the browser. Additionally with port forwarding for HTTPS not working I am then unable to use the Port Privacy feature to make it public to get peer feedback of a .NETCore site running.

Do you know if anyone in the .NETCore team has tried to use some of the dotnet new web templates in Codespaces?
As I assume they might have run into similar teething issues that I seem to be having.


Look forward to making Codespaces a super great place to build and work with .NETCore WebApp development

**Reply by Chuxel** [reply]
@warrenbuckley If you go and update the port in the port forwarding UX to be "https", does it work from the web?

Digging in a bit more, we think **this should actually work if you set the port to HTTPS** in the ports view (or using `portAttributes` in `devcontainer.json`) as you described above.  

What should be happening is that the self-signed certificate is ignored by the forwarding service in "https" mode and its own certificate applied. We don't support using your own certificate on the web - but we should be able to convert it to our own. (And this issue also highlights the need for an https-passthrough variation... though http mostly gets you that.)

This is why VS Code client also redirects to HTTPS when you set the protocol. For an existing project, you could then just do a `dotnet dev-certs https --clean` to generate the needed cert for your config . 

The other way will work for now (if you stick to `http`), but @derekbekoe's team is going to take a look.

**Reply by warrenbuckley** [reply]
>If you go and update the port in the port forwarding UX to be "https", does it work from the web?
>
> Digging in a bit more, we think **this should actually work if you set the port to HTTPS** in the ports view (or using `portAttributes` in `devcontainer.json`) as you described above.

Sadly no, it was the first thing I tried in the hope that it would then just magically start to work. So it sounds like a bug, if you was convinced that this would then work for me.

Let me know if you have any other ideas or solutions for me to try or when you believe the problem has been fixed. Will gladly retry it out @Chuxel and @derekbekoe 😄 


### devcontainer.json
Here is a copy of my devcontainer.json with the explicit port forwarding extra attributes set for the protocol

```json
// For format details, see https://aka.ms/devcontainer.json. For config options, see the README at:
// https://github.com/microsoft/vscode-dev-containers/tree/v0.202.5/containers/dotnet-mssql
{
	"name": "C# (.NET) and MS SQL",
	"dockerComposeFile": "docker-compose.yml",
	"service": "app",
	"workspaceFolder": "/workspace",

	// Set *default* container specific settings.json values on container create.
	"settings": {
		"mssql.connections": [
			{
				"server": "localhost,1433",
				"database": "",
				"authenticationType": "SqlLogin",
				"user": "sa",
				"password": "P@ssw0rd",
				"emptyPasswordInput": false,
				"savePassword": false,
				"profileName": "mssql-container"
			}
		]
	},

	// Add the IDs of extensions you want installed when the container is created.
	"extensions": [
		"ms-dotnettools.csharp",
		"ms-mssql.mssql"
	],
	"remoteEnv": {
		"ASPNETCORE_Kestrel__Certificates__Default__Password": "SecurePwdGoesHere",
		"ASPNETCORE_Kestrel__Certificates__Default__Path": "/home/vscode/.aspnet/https/aspnetapp.pfx",
	},

	// Use 'forwardPorts' to make a list of ports inside the container available locally.
	// "forwardPorts": [5000, 5001],

	// [Optional] To reuse of your local HTTPS dev cert:
	//
	// 1. Export it locally using this command:
	//    * Windows PowerShell:
	//        dotnet dev-certs https --trust; dotnet dev-certs https -ep "$env:USERPROFILE/.aspnet/https/aspnetapp.pfx" -p "SecurePwdGoesHere"
	//    * macOS/Linux terminal:
	//        dotnet dev-certs https --trust; dotnet dev-certs https -ep "${HOME}/.aspnet/https/aspnetapp.pfx" -p "SecurePwdGoesHere"
	// 
	// 2. Uncomment these 'remoteEnv' lines:
	//    "remoteEnv": {
	// 	      "ASPNETCORE_Kestrel__Certificates__Default__Password": "SecurePwdGoesHere",
	//        "ASPNETCORE_Kestrel__Certificates__Default__Path": "/home/vscode/.aspnet/https/aspnetapp.pfx",
	//    },
	//
	// 3. Next, copy your certificate into the container:
	//      1. Start the container
	//      2. Drag ~/.aspnet/https/aspnetapp.pfx into the root of the file explorer
	//      3. Open a terminal in VS Code and run "mkdir -p /home/vscode/.aspnet/https && mv aspnetapp.pfx /home/vscode/.aspnet/https"

	// postCreateCommand.sh parameters: $1=SA password, $2=dacpac path, $3=sql script(s) path
	"postCreateCommand": "bash .devcontainer/mssql/postCreateCommand.sh 'P@ssw0rd' './bin/Debug/' './.devcontainer/mssql/'",
	"features": {
		"github-cli": "latest",
		"azure-cli": "latest"
	},
	"portsAttributes": {
		"5001": {
			"label": "HTTPS",
			"onAutoForward": "notify",
			"protocol": "https",
			"elevateIfNeeded": true
		},
		"5000": {
			"label": "HTTP",
			"onAutoForward": "notify",
			"protocol": "http",
			"elevateIfNeeded": true
		}
	}
}
```

**Reply by mattwoberts** [reply]
I'm having _exactly_ the same issue - I posted this: https://stackoverflow.com/questions/69754444/running-an-asp-net-core-web-app-in-github-codespaces-gives-502-bad-gateway/69827715#69827715

And yes - I tried also swapping port forwarding on port 5001 to use HTTPS instead of the HTTP default with no joy.

I did manage to get it working over HTTP, it's just HTTPS that's causing an issue.

Since I'm not entirely familiar with the setup of how the port forwarding ends up inside the docker container, I was wondering if it's some sort of issue with nginx not being able to use a valid SSL certificate, and thus causing the 502, but I don't know where to look for logs or anything to diagnose this any further. 

I did try to follow the steps in the devcontainer.json fiile to reuse the local https dev cert (I wasn't sure if this was a pre-req for using HTTPS) but that didn't make a difference either.

**Reply by Chuxel** [reply]
Yep. This did work - there's some sort of bug here. Temporary workaround is to use http and localhost, but ngeix should be consuming the self signed cert so that the web HTTPS URI gets the appropriate Codespaces cert for its domain. You shouldn't need to deal with copying certs around.

What we don't yet support is reusing an organization's cert for the last mile through the web - the codespaces cert is always used. However, for those cases, you can use VS Code client with the default port protocol type (which is also used when http is specified). This is a pure TCP socket when forwarded to localhost and thus would support any protocol. That said, this is a different scenario than that.

**Reply by warrenbuckley** [reply]
Ok so if I understand correctly then this is a bug and HTTPS will work with CodeSpaces web port forwarding url?

Is there any issue to follow along to get updates or insights so I know when to try this out again?

**Reply by Chuxel** [reply]
Correct - This is the right spot to watch!

**Reply by warrenbuckley** [reply]
@Chuxel any updates or feedback on this from you and the relevant teams ?

**Reply by Chuxel** [reply]
This should be resolved!  Change the protocol for port that is forwarded HTTPS and it will translate.  Either right click and update the port in the ports view in VS Code, or use the `portsAttributes` property in devcontainer.json.

```json
"portsAttributes": {
    "7062": {
        "protocol": "https"
    }
}
```

Note that your app may also do redirects that hard code localhost, so if hitting the http port (rather than the https port) is resulting in an error, that's likely what is going on.  Forwarded ports follow the form `${CODESPACE_NAME}-<port>.githubpreview.dev` (where `CODESPACE_NAME` is an environment variable in the codespace) on port 443 and will be used in https mode from local VS Code client as well.

**Reply by micya** [reply]
Came across this today. Go to "Ports" tab, right click on port, select "Change port protocol", and select "HTTPS".

![ports](https://user-images.githubusercontent.com/53619397/148658336-17a0f294-e93a-45e4-8c92-5ec2442f2119.png)

**Comment by joshjohanning** [comment]
We were able to get ours to load after 
1) manually flipping port 5001 to https - it was set to http
2) waiting like a minute

Then it came up

**Reply by Scryptium-DamonB** [reply]
Thank you for mentioning waiting like a minute! Mine was setup with a different port. Changed the port protocol from http to https, and this was key: waited (5 minutes while I answered the front door), tried and it worked! I had flipped the port protocol earlier, but didn't wait for a bit before trying again. Good to know for anyone new to this way of developing software.

**Comment by skavor** [comment]
Delete this line if you are using .net core 6. it has worked for me 
![Capture d’écran 2022-11-22 101129](https://user-images.githubusercontent.com/35343823/203273494-b3a37ac3-7688-4f21-b1a8-43990b173038.png)

**Reply by joshjohanning** [reply]
This is what I ended up doing! Rely on server-side HTTPS redirection (like in azure app settings) instead of putting in the app.

**Comment by AgileMindExec** [comment]
I was working on this project on Codespaces and got the same error: https://github.com/LinkedInLearning/hands-on-aspnet-razor-pages-3084715

Commenting this line number 16 helped solve the issue.

![image](https://github.com/community/community/assets/59906964/b9ccd688-b519-4368-b718-d6140e9a2b9a)

**Reply by rajbos** [reply]
Thanks for adding that @AgileMindExec! It will be helpful for the next person!

I would suspect then that you also can check the forwarded ports setting in the devContainer.json file. Maybe that is not configured for https then?

**Reply by AgileMindExec** [reply]
That's weird because it was set to https! I tried switching both before commenting it out and neither of those settings worked.
